Editor additions for MipMaps and rd_textures

DDarby-Lewis opened the pull request:
This PR adds mipmap selector for texture previews. The mipmap selector only shows if there is at least one mipmap level.

<img width="986" height="571" alt="image" src="https://github.com/user-attachments/assets/c8c6acbb-ee79-4522-8b54-3d57f7a29136" />
<img width="991" height="605" alt="image" src="https://github.com/user-attachments/assets/5550bcb3-9712-4d42-a049-cdaf1e5da591" />

It also enables texture previews for rd_textures. 
Finally there is a small addition for rd_textures using RenderingDevice.DATA_FORMAT_R8G8B8A8_UNORM format to properly create their SRGB view - bringing them inline with the same assumptions used by other textures with the R8G8B8A8 format that they will be in SRGB. 

The project this is being developed for utilises texture generation with compute textures.

I would also like to consider if it would be possible/reasonable to make the `TexturePreview` class accessible from GDScript so that my procedural texture class could use it without having to write that class into the core engine?

*Bugsquad edit:* Closes https://github.com/godotengine/godot-proposals/issues/12873

Ivorforce commented:
Hello, thank you for contributing to Godot engine! This looks like an interesting addition :)

New feature PRs should be linked to a [feature proposal](https://github.com/godotengine/godot-proposals/issues). I quickly searched through and found nothing of the sort, so you might have to open one yourself.

DDarby-Lewis commented:
I would appreciate any thoughts or comments on this, it's my first PR (and second contribution) for Godot.
 
Something else to consider here is if I could add some more generic way to make the texture preview applicable, since would like to move my procedural texture class from using Texture2DRD to being based on Texture2D and handling the other parts myself.
Reason being that sometimes I would need to be able to switch from using a rd texture and other times a texture 2d as a the rendering server side, as decals for example don't support rd textures. 
My current solution is to have two separate classes, with separate bases, but that is messy since I cannot switch a texture from one to the other easily.
I was thinking perhaps to use utilise a some general logic like if the class is Texture2D, and it has functions for `_get_mipmap_count` and `_get_format` (inline with the existing override functions for Texture2D class) to provide the additional data that the texture preview is after then it will apply a preview.

Right now the texture preview class contains a lot of logic that is repeated for various texture subclasses, and it would be better OOP if that logic lived on the subclasses (perhaps even on an override function), each class could even mark itself with a const if it wanted texture preview to apply. This would both DRY up the TexturePreview class and move the logic to the responsible subclasses. Thoughts?

If there is an appetite for a feature that adds a generic procedural texture class to the core engine I can look at converting my GDscript work into c++ after this,  but that would be good chunk of work since it uses a custom set of compute pipeline resources along side some.

DDarby-Lewis commented:
@Ivorforce I have added a feature proposal here: https://github.com/godotengine/godot-proposals/issues/12873

beicause commented on servers/rendering/renderer_rd/storage_rd/texture_storage.cpp:
Is this change needed? What would happen without it?

DDarby-Lewis replied:
This is a pre-existing issues with rd textures but is more of an issue now that rd textures are supported in the texture preview, in that the preview is 2D and if your texture is for 3D it will be linear, so it displays wrong without this change.
You can see the mentioned issue for screenshots.
I figure that since it is connected to related functionality in the TexturePreview it makes sense to include both.

beicause replied:
I didn't test it, but I'm still skeptical about this.

`texture_create_shared` is necessary, otherwise, you might accidentally release the `rd_texture`, and it might wrongly assign a srgb format texture to `rd_texture`.

Note that srgb format textures are generally used in 3D for samplers to automatically perform sRGB → linear conversion. In 2D you should use non-srgb formats.

beicause replied:
What 8 bit texture do you use? R8G8B8A8 should work fine, but R8 or R8G8 or R8G8B8 won't, their srgb versions' device support is not good.

DDarby-Lewis replied:
Thank you for testing - it's always good to have a second set of eyes confirm :) 
I'll look at creating a separate PR.

DDarby-Lewis replied:
I have taken a look at the PR you have raised and we have had slightly different tacts - I haven't added the shared formats under the assumptions that it would prevent any disruption to work already relying on the current way that it works and instead I only create the srgb view if the associated shared format is already present (again on the assumption that if anyone has included theses then they are expecting the to be used).

I then include the following lines in my procedural texture generator in my project:
```
    if _format == RenderingDevice.DATA_FORMAT_R8G8B8A8_UNORM:
        format.add_shareable_format(_format)
        format.add_shareable_format(RenderingDevice.DATA_FORMAT_R8G8B8A8_SRGB)
``` 

Do we need to have a discussion on the best way to approach this? If the fact that it is missing is purely bugged behaviour then I'd agree to just adding the formats. However if, as you are pointing out in your PR, it may lack full support, perhaps making the approach of only adding a view when required preferable?

DDarby-Lewis commented:
> I support this, the ability to visualize mipmaps is useful in many cases. Could the selection settings be changed to a horizontal slider, though? I feel as if that would make it easier to use.

I looked at this but on the scale of this UI element a slider is fiddly and the selection boxes show clearly the mipmap number as well as matching the visuals for the channel selections.

DDarby-Lewis commented:
I will address the linting issues tomorrow and repush, from what I understand I should undo my commit and force push a single commit with the changes, is that correct?

DDarby-Lewis commented:
I also don't have a good icon for the mipmap toggle, if anyone can suggest something that'd be great!

LiveTrower commented:
I think it would be better to do this with a SpinBox:

https://github.com/user-attachments/assets/c2ddbb1d-f927-4139-a6ea-b4f53d1bea9e

I also see that there's a missing implementation for mipmap pre-visualization for other texture types that support mipmaps, such as Texture3D, CubeMap, Texture2DArray, etc.

One thing I think would also be good is to add a project or editor setting that allows switching between linear or nearest filtering, as mipmaps are normally used with linear filtering.

DDarby-Lewis commented:
@LiveTrower I feel that the radio button selectors are both easier to use and nicer looking, but I'm happy to be outvoted and adjust it 😄 

As for mipmap visuals on other texture types, I have here only added a mipmap option to textures in the TexturePreview that support retrival of the number of mipmaps - I think it would perhaps make sense to expand the Texture class (or subclasses) with an implementation of the `get_format` function (with assisiated override `_get_format`) and a `get_mipmap_count` (and assositated _ override) as suggested in the feature proposal here: https://github.com/godotengine/godot-proposals/issues/12873
That way we could enable the texture preview more widely and rely on those functions without adding custom logic per class.
However, that expansion seems out of scope for this change to add the initial functionality.
Perhaps if you support it or have thoughts you could add them to that discussion?

DDarby-Lewis commented:
So along with the addition of the virtual _get_image function this allows the texture preview to apply to any texture with a working get_image function - I have disabled for any class called `GradientTexture.*` for obvious reasons. If any has any other textures to ignore here please let me know. I'd like this to get merged soon so if anyone has final comments please add them and then I'll  split into two MRs as suggested.

MR 1: Editor additions 
MR 2: core function enhancements

DDarby-Lewis commented:
Some how hit a compilation failure in CI that I don't get locally - should be fixed now.
I'd appreciate help testing this if anyone is willing to pull and build to check any edge cases I may have missed since this affects a lot of things with Texture2D.

AThousandShips commented:
To match you need to enable the same build options as the engine does, specifically `dev_mode=yes`

DDarby-Lewis commented:
@AThousandShips thanks - I was using `dev=yes` not dev_mode - I also forgot to do a docs update. Hopefully will pass CI now, still waiting on some other people to test it then I'll split it out.

Calinou commented:
> So it failed again, I'm not really sure what it needs at this point though so any advice would be appreciated.

CI is failing because `get_format()` is no longer exposed in the ImageTexture and PortableCompressedTexture2D classes, even though it's now exposed in Texture2D instead (both ImageTexture and PortableCompressedTexture2D inherit from Texture2D). In terms of compatibility, this is acceptable for scripts, but GDExtensions will break. As I understand it, to preserve compatibility with existing projects, you'd need to expose dedicated `get_format()` method overrides in ImageTexture and PortableCompressedTexture2D.

cc @dsnopek

Calinou commented:
Functionality-wise, this seems good, but textures with many mipmaps (4096×4096 or 8192×8192) will cause the inspector to overflow and become wider:

<img width="404" height="300" alt="image" src="https://github.com/user-attachments/assets/12a68d5a-3eb2-472c-ad18-02e79b0d8bc9" />

I would probably reuse the same UI as Texture3D layers do, which uses a single SpinBox to select the displayed layer. (You can also drag the mouse over its up/down arrows to change between layers with mouse motion.)

arkology commented:
Also maybe it makes sense to put this above color channel selector as more common functionality (but in general it's a matter of taste).

DDarby-Lewis commented:
<img width="1032" height="794" alt="image" src="https://github.com/user-attachments/assets/b57bef26-2f69-4e72-930c-37358e4eeb13" />


Edited to use a spin box.

And I *think* I have done what is required to address the expected API changes.

Calinou reviewed (commented):
Works great now :slightly_smiling_face: 

My only remaining concern is that the layer selector is in the top-left corner, while it's in the top-right corner for Texture2DArray:
 
Texture2D | Texture2DArray
-|-
<img width="279" height="276" alt="image" src="https://github.com/user-attachments/assets/186441c3-e32b-4943-ae00-2b35f149b213" /> | <img width="279" height="276" alt="image" src="https://github.com/user-attachments/assets/40f59581-ded8-42e4-9c06-6b638f7d1bc9" />

I suggest moving the layer selector added in this PR to the top-right corner to match Texture2DArray. This also allows the channel selector to remain in the same position for both types.

DDarby-Lewis commented:
@Calinou thanks for catching that. I'll take a look at it.
Just want to flag that my custom Texture2D derived gd script (what I have used for testing this) has issues when trying to "make unique" and it seems to delete the resource. I don't think that is related to this MR, I'll try and test to confirm.

LiveTrower commented:
I don't want to be a pain, but I thought the toggle button that showed or hid this option was a thousand times better. The fact that it's out in the open takes up a lot of space because the texture preview is too small. I know you can increase the panel's size horizontally, but this will clash and be confusing when you want to add this same option to other texture types, like in the Texture2DArray example. As you can see, it will end up looking something like this:
<img width="279" height="276" alt="preview" src="https://github.com/user-attachments/assets/f3c031d6-e409-4d79-a3d2-36c928c06f65" />

A simple solution is to add a tooltip that explains what the spinbox does when you hover over it. The toggle button is another option that would solve both the space issue and the confusion about what it does.

arkology commented:
I think sizing issue could be solved in future with ability to hide everything using special context menu.
Talking about tooltip, this will be useful for sure.

KoBeWi commented on editor/scene/texture/texture_editor_plugin.h:
```diff
@@ -33,13 +33,15 @@
 #include "editor/inspector/editor_inspector.h"
 #include "editor/plugins/editor_plugin.h"
 #include "scene/gui/margin_container.h"
+#include "scene/gui/spin_box.h"
```
This should be in cpp file, and SpinBox forward-declared instead.

DDarby-Lewis replied:
Done

KoBeWi commented on editor/scene/texture/texture_editor_plugin.cpp:
```diff
@@ -128,6 +130,11 @@ static Image::Format get_texture_2d_format(const Ref<Texture2D> &p_texture) {
 	}
 
+	const Ref<Texture2DRD> rd_texture = p_texture;
+	if (rd_texture.is_valid() && RD::get_singleton()->texture_owner.get_or_null(rd_texture->get_texture_rd_rid()) != nullptr) {
```
Is the RD check needed? Can the texture have valid reference, but invalid data?

beicause replied:
Should use `RD::get_singleton()->texture_is_valid` instead.

DDarby-Lewis replied:
That is basically the exact same check - have you tested it?

KoBeWi commented on editor/scene/texture/texture_editor_plugin.cpp:
```diff
@@ -306,7 +349,26 @@ TexturePreview::TexturePreview(Ref<Texture2D> p_texture, bool p_show_metadata) {
 
 bool EditorInspectorPluginTexture::can_handle(Object *p_object) {
-	return Object::cast_to<ImageTexture>(p_object) != nullptr || Object::cast_to<AtlasTexture>(p_object) != nullptr || Object::cast_to<CompressedTexture2D>(p_object) != nullptr || Object::cast_to<PortableCompressedTexture2D>(p_object) != nullptr || Object::cast_to<AnimatedTexture>(p_object) != nullptr || Object::cast_to<Image>(p_object) != nullptr;
+	if (p_object->get_class().begins_with("GradientTexture")) {
```
```suggestion
	if (Object::cast_to<GradientTexture1D>(p_object) || Object::cast_to<GradientTexture2D>(p_object)) {
```
It's better to avoid string checks.

KoBeWi commented on editor/scene/texture/texture_editor_plugin.cpp:
```diff
@@ -247,6 +274,7 @@ TexturePreview::TexturePreview(Ref<Texture2D> p_texture, bool p_show_metadata) {
 		material.instantiate();
 		material->set_shader(shader);
 		material->set_shader_parameter("u_channel_factors", Vector4(1, 1, 1, 1));
+		material->set_shader_parameter("lod", 0.0);
```
Setting the default shouldn't be necessary.

DDarby-Lewis commented:
@KoBeWi thank you for the review, I will try and update tomorrow.

DDarby-Lewis commented:
<img width="867" height="424" alt="image" src="https://github.com/user-attachments/assets/de2231ad-3879-472d-8176-befdb6a09050" />
Pushed a update to all requested changes

DDarby-Lewis commented on doc/classes/Texture2D.xml:
```diff
@@ -113,6 +133,11 @@
 				Draws a part of the texture using a [CanvasItem] with the [RenderingServer] API.
 			</description>
 		</method>
+		<method name="get_format" qualifiers="const">
```
Done

KoBeWi commented on editor/scene/texture/texture_editor_plugin.cpp:
```diff
@@ -271,6 +298,22 @@ TexturePreview::TexturePreview(Ref<Texture2D> p_texture, bool p_show_metadata) {
+	const int mipmaps = get_texture_mipmaps_count(p_texture);
+	if (mipmaps > 0) {
+		mipmap_spinbox = memnew(SpinBox);
+		mipmap_spinbox->set_tooltip_text("Mipmap level viewer.");
```
```suggestion
		mipmap_spinbox->set_tooltip_text(TTRC("Mipmap level viewer."));
```
Also I'd change the text to something like `Mipmap index.`. It's odd calling SpinBox a "viewer".

KoBeWi reviewed (approved):
Looks good on editor side.

DDarby-Lewis commented:
@KoBeWi Tests show it has failed loading an SVG on macOS version, no clue why as it works on other builds. I don't have access to a mac to test this, is there any chance someone else can take a look?

bruvzg commented:
> test_image.h:162: ERROR: CHECK( image_svg->load_svg_from_buffer(data_svg) == OK ) is NOT correct!

The same error shows in CI time-to-time, I was never able to reproduce it locally. Most likely it's not related to the PR, so just restart the failed task.

DDarby-Lewis commented:
Great, thanks @bruvzg

AThousandShips commented on doc/classes/Texture2D.xml:
```diff
@@ -145,5 +180,11 @@
+		<method name="has_mipmaps" qualifiers="const">
+			<return type="bool" />
+			<description>
+				Returns true if the texture has mipmaps.
```
```suggestion
				Returns [code]true[/code] if the texture has mipmaps.
```

AThousandShips commented on doc/classes/Texture2D.xml:
```diff
@@ -66,6 +83,12 @@
+		<method name="_has_mipmaps" qualifiers="virtual const">
+			<return type="bool" />
+			<description>
+				Called when the [Texture2D]'s has_mipmaps is called.
```
```suggestion
				Called when [method has_mipmaps] is called.
```

AThousandShips commented on doc/classes/Texture2D.xml:
```diff
@@ -48,12 +48,29 @@
+		<method name="_get_mipmap_count" qualifiers="virtual const">
+			<return type="int" />
+			<description>
+				Called when the [Texture2D]'s get_mipmap_count is called.
```
```suggestion
				Called when [method get_mipmap_count] is called.
```

AThousandShips commented on doc/classes/Texture2D.xml:
```diff
@@ -48,12 +48,29 @@
+		<method name="_get_format" qualifiers="virtual const">
+			<return type="int" enum="Image.Format" />
+			<description>
+				Called when the [Texture2D]'s get_format is called.
```
```suggestion
				Called when [method get_format] is called.
```

AThousandShips commented on misc/extension_api_validation/4.4-stable.expected:
```diff
@@ -327,4 +327,12 @@ GH-108825
 Validate extension JSON: Error: Field 'classes/EditorExportPlatformExtension/methods/_get_option_icon/return_value': type changed value in new API, from "ImageTexture" to "Texture2D".
 
+Return type changed to allow returning both ImageTexture and SVGTexture. Compatibility method registered.
 Return type changed to allow returning both ImageTexture and DPITexture. Compatibility method registered.
```
```suggestion
```

AThousandShips commented on doc/classes/Texture2D.xml:
```diff
@@ -48,12 +48,29 @@
 		</method>
+		<method name="_get_image" qualifiers="virtual const">
+			<return type="Image" />
+			<description>
```
Needs a description

DDarby-Lewis commented:
@KoBeWi could you help me resolve whatever issue is causing the Linux editor failure and tell me if there is anything else I can do to get this merged?

DDarby-Lewis commented:
I have rebased on master and hit some errors - mainly from this:
https://github.com/godotengine/godot-proposals/issues/7379

It would appear there are compilation issues, but also potentially the aim of that proposal has overlapped with the reason I started this work.

I still like the option to have a custom Texture2D class return with _get_image.

DDarby-Lewis commented:
Ok - I have updated this now - I have also added missing docs for the drawable texture - not sure how those got left out.
@KoBeWi  @beicause can I get the CI approved?

AThousandShips commented on doc/classes/DrawableTexture2D.xml:
```diff
@@ -38,6 +38,40 @@
+			<return type="void" />
+			<param index="0" name="width" type="int" />
+			<description>
+				set the width of this DrawableTexture.
```
```suggestion
				Sets the width of this DrawableTexture.
```

DDarby-Lewis commented:
I think I've addressed all the issues - however GitHub is having issues - so waiting for those to be resolved.

AThousandShips commented:
Now fixed, you just need to re-apply the documentation changes (you undid them) and fix the format

AThousandShips commented:
It needs more maintainers to review and approve it, it has one already

DDarby-Lewis commented:
I can't see the rules for minimal approvals, do you know how many it requires?

AThousandShips commented:
It depends from case to case, it's not a simple case of rules it's a matter of specific needs of the project

KoBeWi commented on editor/scene/texture/texture_editor_plugin.cpp:
Why this method still exists if `get_format()` is in Texture2D?

KoBeWi reviewed (commented):
The editor changes look fine, but there is one potential crash that needs to be addressed.

KoBeWi commented on editor/scene/texture/texture_editor_plugin.cpp:
```diff
@@ -155,13 +147,20 @@ static int get_texture_mipmaps_count(const Ref<Texture2D> &p_texture) {
 			image = atlas->get_image();
 		}
+	} else if (rd_texture.is_valid()) {
+		if (RD::get_singleton()->texture_is_valid(rd_texture->get_texture_rd_rid())) {
```
Also needs safeguard
```suggestion
		if (RD::get_singleton() && RD::get_singleton()->texture_is_valid(rd_texture->get_texture_rd_rid())) {
```

KoBeWi reviewed (approved):
The editor changes look good.

Calinou commented:
This fails to compile for me on Linux/GCC when rebased on top of `master`:

```
Generating editor/themes/builtin_fonts.gen.h ...
In file included from editor/scene/texture/.scu/scu_editor_scene_texture.gen.cpp:5:
./editor/scene/texture/texture_editor_plugin.cpp: In function 'Image::Format get_texture_2d_format(const Ref<Texture2D>&)':
./editor/scene/texture/texture_editor_plugin.cpp:137:38: error: 'RD' has not been declared; did you mean 'RID'?
  137 |         if (rd_texture.is_valid() && RD::get_singleton() && RD::get_singleton()->texture_is_valid(rd_texture->get_texture_rd_rid())) {
      |                                      ^~
      |                                      RID
./editor/scene/texture/texture_editor_plugin.cpp:137:61: error: 'RD' has not been declared; did you mean 'RID'?
  137 |         if (rd_texture.is_valid() && RD::get_singleton() && RD::get_singleton()->texture_is_valid(rd_texture->get_texture_rd_rid())) {
      |                                                             ^~
      |                                                             RID
./editor/scene/texture/texture_editor_plugin.cpp: In function 'int get_texture_mipmaps_count(const Ref<Texture2D>&)':
./editor/scene/texture/texture_editor_plugin.cpp:160:21: error: 'RD' has not been declared; did you mean 'RID'?
  160 |                 if (RD::get_singleton() && RD::get_singleton()->texture_is_valid(rd_texture->get_texture_rd_rid())) {
      |                     ^~
      |                     RID
./editor/scene/texture/texture_editor_plugin.cpp:160:44: error: 'RD' has not been declared; did you mean 'RID'?
  160 |                 if (RD::get_singleton() && RD::get_singleton()->texture_is_valid(rd_texture->get_texture_rd_rid())) {
      |                                            ^~
      |                                            RID
```

I have tried a fresh build with no cache just in case, and it still occurs.

DDarby-Lewis committed:
Editor additions for MipMaps and rd_textures

DDarby-Lewis commented:
@Calinou I have pushed a fix for that. Would love to merge this so I can stop having to patch fixes in from changes to main ;)

Calinou commented on editor/scene/texture/texture_editor_plugin.cpp:
```diff
@@ -277,6 +290,21 @@ TexturePreview::TexturePreview(Ref<Texture2D> p_texture, bool p_show_metadata) {
 	const Image::Format format = p_texture.is_valid() ? get_texture_2d_format(p_texture.ptr()) : Image::FORMAT_MAX;
 	const uint32_t components_mask = format != Image::FORMAT_MAX ? Image::get_format_component_mask(format) : 0xf;
 
+	// Setup Mipmap selector
```
```suggestion
	// Setup Mipmap selector.
```

Calinou reviewed (approved):
Tested locally, it works as expected. Code looks good to me.

I agree we should implement mipmap previewing support to types other than Texture2D (that is: Texture2DArray, Texture3D, Cubemap, and CubemapArray), but this can be added in a future PR. Having support for mipmap previewing on Texture2D will already be pretty useful for VFX use cases, as well as troubleshooting procedural texture generation with the Image class.
